CDK SST blog (#12592)
pulumi/docs: content/blog/from-cdk-pulumi-evolution-of-sst/index.md

# https://www.linkedin.com/feed/update/urn:li:activity:7169021002394296320
# https://www.linkedin.com/feed/update/urn:li:activity:7155606616455737345
# https://twitter.com/PulumiCorp/status/1763265391042654623
# https://twitter.com/PulumiCorp/status/1762900472489185492
# https://twitter.com/PulumiCorp/status/1755637618631405655

social:
    twitter: Curious about SST's evolution from CDK to Pulumi? Discover how this transition is empowering developers with a more flexible, provider-agnostic approach to infrastructure management.
    linkedin: The evolution of cloud computing has reshaped how developers interact with infrastructure. For SST, a framework built to empower application developers, the journey began with AWS CDK but soon encountered limitations, and the search for a more flexible, provider-agnostic solution led them to Pulumi. Discover how Pulumi is redefining infrastructure management and what this means for the future of SST. Dive into the full story at [link].

# See the blogging docs at https://github.com/pulumi/docs/blob/master/BLOGGING.md
# for details, and please remove these comments before submitting for review.
---

> As cloud computing continues to evolve, so do the tools and frameworks that developers rely on to manage their infrastructure. For Serverless Stack (SST), a framework built to empower application developers, the journey began with AWS's CDK. While CDK offered a way to define infrastructure using familiar programming languages, it soon became clear that more was needed to meet developers' growing needs.

Reviewer: Suggestion:

> For Serverless Stack (SST), a framework built

->

> For the team building [Serverless Stack (SST)](https://sst.dev/), a framework built

It was during this time that the SST team discovered Pulumi, a modern IaC platform that takes a fundamentally different approach to infrastructure management. Instead of generating an intermediary format, Pulumi treats the infrastructure code as a first-class program executed directly during deployment.

This paradigm shift had several important implications for SST and its users:

- **Visibility and Extensibility**: With Pulumi, the infrastructure code is the same code that is running during deployment, which means there is much greater visibility into the deployment process and the ability to extend or customize it as needed.
- **Multi-Cloud Capabilities**: Pulumi's provider-agnostic approach allows SST's team to easily work with a wide range of cloud and on-premises platforms, giving their users the flexibility to deploy their infrastructure wherever it makes the most sense for their application.
- **Simplified Mental Model**: For application developers, the Pulumi model of "your code is the deployment" aligns much more closely with their existing mental models and workflows, making it easier for them to adopt and work with IaC tools.

## Transitioning to Pulumi

> Transitioning SST [from CDK to Pulumi](https://www.pulumi.com/tutorials/importing/) was not a trivial undertaking. Still, they knew it was a necessary step to truly fulfill our mission of empowering application developers with powerful infrastructure management capabilities.

Reviewer: > fulfill our mission

->

> fulfill their mission

This flexibility has been particularly valuable for their users, who may have workloads or requirements that span multiple cloud providers or even on-premises infrastructure. With SST built on Pulumi, they can now manage all of their infrastructure through a single, consistent interface without having to juggle multiple tools or approaches.

### Simplified Mental Model

One of Pulumi's most significant benefits for SST users is the simplified mental model it provides. By treating the infrastructure code as a first-class program, Pulumi aligns much more closely with how application developers think and work.

Instead of having to navigate the complexities of intermediary formats, deployment pipelines, and the separation between infrastructure code and deployment code, SST users can now focus on writing their infrastructure logic in the same programming languages they use for their application code. This makes it much easier for them to understand, maintain, and extend their infrastructure as their needs evolve.

## The Future of SST with Pulumi

> With Pulumi's foundation in place, they can now focus on further enhancing the developer experience and expanding the capabilities of our framework. Some of the key areas they are exploring include:

Reviewer: > capabilities of our framework

->

> capabilities of their framework

# the date and time this file was generated. Dates are used for display and
# ordering purposes only; they have no effect on whether or when a post is
# published. To influence the ordering of posts published on the same date, use
# the time portion of the date value; posts are sorted in descending order by
# date/time.
date: 2024-09-25T07:32:40Z

# The draft setting determines whether a post is published. Set it to true if
# you want to be able to merge the post without publishing it.
draft: false

# Use the meta_desc property to provide a brief summary (one or two sentences)
# of the content of the post, which is useful for targeting search results or
# social-media previews. This field is required or the build will fail the
# linter test. Max length is 160 characters.
meta_desc: Explore how SST evolved from AWS CDK to Pulumi, overcoming limitations to offer a flexible, provider-agnostic infrastructure management solution for developers.

# The meta_image appears in social-media previews and on the blog home page. A
# placeholder image representing the recommended format, dimensions and aspect
# ratio has been provided for you.

> meta_image: meta.png

Author: ```suggestion
meta_image: sst-with-pulumi-infrastructure-as-code-deployments.png
```

> meta_image: sst-with-pulumi-infrastructure-as-code-deployments.png

Author: ```suggestion
meta_image: "sst-with-pulumi-infrastructure-as-code-deployments.png"
```

- **Improved Deployment Workflows**: Building on Pulumi's extensibility, they can create [more advanced deployment workflows](https://www.pulumi.com/docs/pulumi-cloud/deployments/) that address the specific needs of application developers, such as faster deployment times, better rollback capabilities, and more granular control over the deployment process.
- **Enhanced Observability and Monitoring**: By treating the infrastructure code as a first-class program, they can provide their users with [better visibility](https://www.pulumi.com/product/pulumi-insights/) into the deployment process and more robust monitoring and observability capabilities, helping them to identify and resolve issues quickly.

As the SST team continues to evolve with Pulumi at its core, they aim to deliver an even more powerful and user-friendly infrastructure management experience for developers. This will empower teams to focus on building great applications while effortlessly managing the underlying infrastructure.

## Conclusion

If you're interested in exploring Pulumi further, here are several ways to get involved:

> - Discover how Pulumi Crosswalk for AWS simplifies “day one” tasks in our [AWS guide](https://www.pulumi.com/docs/clouds/aws/guides/).
> - Join an upcoming Platform Engineering & DevOps in-person meetup, [find a location near you](https://info.pulumi.com/platform-engineering-devops-series)
> - Watch our on-demand workshop [Getting Stated with Infrastructure as Code on AWS](https://www.pulumi.com/resources/getting-started-with-iac-pulumi-aws/)
> - Register for one of our upcoming [Platform Engineering or DevOps workshops](https://www.pulumi.com/resources/#upcoming).

Author: ```suggestion
- Discover how Pulumi Crosswalk for AWS simplifies “day one” tasks in our [AWS guide](https://www.pulumi.com/docs/clouds/aws/guides/)
- Join an upcoming Platform Engineering & DevOps in-person meetup, [find a location near you](https://info.pulumi.com/platform-engineering-devops-series)
- Watch our on-demand workshop [Getting Stated with Infrastructure as Code on AWS](https://www.pulumi.com/resources/getting-started-with-iac-pulumi-aws/)
- Register for one of our upcoming [Platform Engineering or DevOps workshops](https://www.pulumi.com/resources/#upcoming)
```

> - Discover how Pulumi Crosswalk for AWS simplifies “day one” tasks in our [AWS guide](https://www.pulumi.com/docs/clouds/aws/guides/)
> - Join an upcoming Platform Engineering & DevOps in-person meetup, [find a location near you](https://info.pulumi.com/platform-engineering-devops-series)
> - Watch our on-demand workshop [Getting Stated with Infrastructure as Code on AWS](https://www.pulumi.com/resources/getting-started-with-iac-pulumi-aws/)
> - Register for one of our upcoming [Platform Engineering or DevOps workshops](https://www.pulumi.com/resources/#upcoming)

Author: ```suggestion
- Discover how Pulumi Crosswalk for AWS simplifies “day one” tasks in our [AWS guide](https://www.pulumi.com/docs/iac/clouds/aws/guides/)
- Join an upcoming Platform Engineering & DevOps in-person meetup, [find a location near you](https://info.pulumi.com/platform-engineering-devops-series)
- Watch our on-demand workshop [Getting Stated with Infrastructure as Code on AWS](https://www.pulumi.com/resources/getting-started-with-iac-pulumi-aws/)
- Register for one of our upcoming [Platform Engineering or DevOps workshops](https://www.pulumi.com/resources/#upcoming)
```

  
---

## Frequently Asked Questions

### What is SST?

SST is a framework that makes building modern full-stack applications on your infrastructure easy.

### What is SST's Ion?

> Ion is the code name for a new engine for deploying SST applications. The constructs (or components) are defined using Terraform providers and deployed using Pulumi instead of CDK and CloudFormation (CFN). Once Ion is stable, it will be released as SST v3.

Reviewer 2: ```suggestion
Ion is the code name for a new engine for deploying SST applications. The constructs (or components) are deployed using Pulumi instead of CDK and CloudFormation (CFN). Once Ion is stable, it will be released as SST v3.
```

> ### Does SST use mostly Terraform or Pulumi?

Reviewer 2: i'd just delete this section ... in general, let's just highlight how great P is

> SST leverages Pulumi behind the scenes as the deployment engine and for provider integrations. Terraform’s providers are bridged through Pulumi.

Author: ```suggestion
SST leverages Pulumi behind the scenes for its providers and deployment engine while also bridging Terraform providers through Pulumi. It only depends on the open-source components of both Pulumi and Terraform.
```

> SST leverages Pulumi behind the scenes for its providers and deployment engine while also bridging Terraform providers through Pulumi. It only depends on the open-source components of both Pulumi and Terraform.

Author: ```suggestion
SST leverages Pulumi behind the scenes for its providers and deployment engine while also bridging Terraform providers through Pulumi.
```

### How does SST make money?

> Although SST is open-sce and free, it also offers the Console, a SaaS service that can automatically deploy y apps and monitor for issues. The Console is optional and includes a free tier.

Author: ```suggestion
SST (Serverless Stack) makes money primarily through its managed service, called SST Console. The Console is optional and includes a free tier. In short, SST primarily relies on SaaS (Software as a Service) revenue through its managed platform and potentially supplemental income streams like enterprise services and consulting.
```

> Although SST is open-sce and free, it also offers the Console, a SaaS service that can automatically deploy y apps and monitor for issues. The Console is optional and includes a free tier.

Reviewer 2: typos
